text/0000-hidden-impls.md

```diff
@@ -0,0 +1,669 @@
+which are available for public use, handle these modifiers,
+or emit an error notifying the user that modifiers are not supported.
+
+## In relation to trait objects
+
+One place where crate B is allowed to use a hidden implementation of crate A
+is when it comes to trait objects. If A provides access to a trait object
+`Box<dyn Property>` (which implies that `Property` is object safe) by returning
+a value of that type, the compiler has no means in the general case to perform
+static analysis on that object to see where the implementation came from.
+This is fine. In this case, it is crate A that expressly gave access to an
+implementation of `Property` for `Thing`. Crate A is also under no obligation
+under the semver rules to provide an erased `Thing` value. As long as the value
+returned behaves observably in an equivalent manner, for some semantic
+definition of "equivalent", no breakage has occurred.
+
+## Recommendation: When you should use hidden implementations
+
+Simply put, it is recommended that your implementation be hidden if you are
+not yet ready to guarantee that the implementation exists to users of your
+API, be they internal APIs within a project, or the public APIs of a crate.
+In most cases, you implementations will not need to be hidden.
+
+## Teaching beginners about hidden implementations
+
+The contents of this RFC are mostly considered to be *advanced* features of Rust.
+However, it should mostly be sufficient to mention to beginners in sections about
+trait implementations that they can hidden, with one or two simple examples.
+The still more advanced details in this RFC can be deferred to the reference.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Grammar and parsing
+
+### Visibility modifier `priv`
+
++ The `PRIV` token will lex the terminal `"priv"`.
+
++ To the production `visibility` we add:
+
+  ```
+  visibility : PRIV | .. ;
+  ```
+
+  where `..` is the old definition of `visibility`.
+
++ The macro fragment specifier `vis` will accept `PRIV`.
+
++ `priv` is considered the default visibility for all contexts but `impl`.
+
+### Visibility modifiers on `impl`
+
++ The grammar of `item_impl` is amended such that `visibility` is optionally
+permitted on *trait implementations*. `visibility` is not allowed before
+inherent implementations.
+
++ `pub` is considered the default visibility for `impl`.
+
+#### `unsafe` and visibility
+
++ When specifying `unsafe` on an `impl`, if a visibility modifier is specified,
+  it must come first. That is, you must write:
+
+  ```rust
+  crate unsafe impl TrustedLen for Bar { .. }
+  ```
+  
+  but never:
+  
+  ```rust
+  unsafe crate impl TrustedLen for Bar { .. }
+  ```
+
+#### Linting on default visibility
+
++ If the *default visibility* for some *context* is specified in that context
+  *directly* in a source file, then a warn-by-default lint `redundant_vis`
+  is raised.
```

Comment:
Hm, so we hard-lint currently on pub in the the below impl; in a way, I could see this being a way to define the "extensible traits" talked about on Discord -- we could say that you could make certain methods of a trait not world-visible. I guess you can also do that with two (or more) traits, but this seems like a reasonable extension of this RFC.

```
trait Foo {
    pub fn bar() {}
}
```

Reply:
Sure; you could have something like:

```rust
// crate A

pub trait Foo {
    crate fn bar();
    pub fn baz();
}
```

and since only crate A can provide a definition for `bar`, then only crate A can provide implementations for `Foo`.

I'll leave this out for now tho :)